language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Function.hs

```diff
@@ -92,3 +95,132 @@ getBuiltinFix = do
+natTrans f g = freshTyName () "Q" >>= \q -> TyForall () q (Type ()) <$> trans f g (TyVar () q)
+
+-- | A type that looks like a natural transformation to Id.
+--
+-- > natTransId F : forall Q :: * . F Q -> Q
+natTransId :: Type TyName () -> Quote (Type TyName ())
+natTransId f = do
+    q <- freshTyName () "Q"
+    pure $ TyForall () q (Type ()) (TyFun () (TyApp () f (TyVar () q)) (TyVar () q))
+
+-- | The 'fixBy' combinator.
+--
+-- fixBy :
+--     forall F :: (* -> *) .
+--     ((forall Q :: * . F Q -> Q) -> (forall Q :: * . F Q -> Q)) ->
+--     ((forall Q :: * . F Q -> F Q) -> (forall Q :: * . F Q -> Q))
+getBuiltinFixBy :: Quote (Term TyName Name ())
+getBuiltinFixBy = do
+    f <- freshTyName () "F"
+    by <- freshName () "by"
+    byTy <- do
+        nt1 <- natTransId (TyVar () f)
+        nt2 <- natTransId (TyVar () f)
+        pure $ TyFun () nt1 nt2
+
+    instantiatedFix <- do
+        fix <- getBuiltinFix
+        nt1 <- natTrans (TyVar () f) (TyVar () f)
+        nt2 <- natTransId (TyVar () f)
+        pure $ TyInst () (TyInst () fix nt1) nt2
+
+    recc <- freshName () "rec"
+    reccTy <- do
+      nt <- natTrans (TyVar () f) (TyVar () f)
+      nt2 <- natTransId (TyVar () f)
+      pure $ TyFun () nt nt2
+
+    h <- freshName () "h"
+    hty <- natTrans (TyVar () f) (TyVar () f)
+
+    r <- freshTyName () "R"
+    fr <- freshName () "fr"
+    let frTy = TyApp () (TyVar () f) (TyVar () r)
+
+    q <- freshTyName () "Q"
+    fq <- freshName () "fq"
+    let fqTy = TyApp () (TyVar () f) (TyVar () q)
+    pure $
+        TyAbs () f (KindArrow () (Type ()) (Type ())) $
+        LamAbs () by byTy $
+        Apply () instantiatedFix $
+        LamAbs () recc reccTy $
+        LamAbs () h hty $
+        TyAbs () r (Type ()) $
+        LamAbs () fr frTy $
+        Apply () (TyInst () (
+                         Apply () (Var () by) $
+                          TyAbs () q (Type ()) $
+                          LamAbs () fq fqTy $
+                          Apply () (TyInst () (Apply () (Var () recc) (Var () h)) (TyVar () q)) $
+                          Apply () (TyInst () (Var () h) (TyVar () q)) (Var () fq))
+                     (TyVar () r))
+        (Var () fr)
+
+-- | Make a @n@-ary fixpoint combinator.
+--
+-- > getBuiltinFixN n :
+--       forall A1 B1 ... An Bn :: * .
+--       forall Q :: * .
+--           ((A1 -> B1) -> ... -> (An -> Bn) -> Q) ->
+--           (A1 -> B1) ->
+--           ... ->
+--           (An -> Bn) ->
+--           Q
+--       forall R :: * . (A1 -> B1) -> ... (An -> Bn) -> R
+getBuiltinFixN :: Int -> Quote (Term TyName Name ())
+getBuiltinFixN n = do
+    asbs <- replicateM n $ do
+        a <- freshTyName () "a"
+        b <- freshTyName () "b"
+        pure (a, b)
+
+    let funTysTo out = foldr (\(a, b) acc -> TyFun () (TyFun () (TyVar () a) (TyVar () b)) acc) out asbs
+
+    fixBy <- getBuiltinFixBy
+
+    x <- freshTyName () "X"
+    s <- freshTyName () "S"
+
+    k <- freshName () "k"
+    kTy <- do
```

Comment:
It would be very good to abstract that out, because various `k`s have simple definitions which are

1. easy to read
2. easy to test
3. very hard to generate inside PLC

and this is just the only place where term-level generation happens, so I think it should be abstracted out and tests should be added (whether in this PR or another).

We may also eventually implement the fully internal version of `fixN` (that would be a good test of the type checker at least) and tests of various `k`s would be very useful.

Reply:
I'm not sure what it is you want to abstract out? The type of `k`, and various values for it?

Reply:
Both. The type is `K F = (F ~> Id) -> F ~> Id` and various instantiations are (mentioning them here just for clarity)

```
λ k f -> f (k λ x y -> x) (k λ x y -> y)
λ k f -> f (k λ x y z -> x) (k λ x y z -> y) (k λ x y z -> z)
...
```

So this doesn't seem to contain any non-normalized types and we can test that.

Reply:
That would be great, but I'm running low on time for today, so I think I'm going to have to leave that as a follow up.

Reply:
That's fine.